visualization/include/pcl/visualization/impl/pcl_visualizer.hpp

```diff
@@ -669,11 +669,26 @@ pcl::visualization::PCLVisualizer::addText3D (
   else
     tid = id;
 
-  std::string id_to_check = viewport > 0 ? tid.append (viewport, '*') : tid;
-  if (contains (id_to_check))
+  if (viewport < 0)
+    return false;
+
+  // check all or an individual viewport for a similar id
+  rens_->InitTraversal ();
+  rens_->GetNextItem (); //discard first because it's not a renderer for the viewps
+  for (size_t i = std::max (viewport, 1); rens_->GetNextItem () != NULL; ++i)
```

Comment:
I'm not sure I understand how this loop works. Could you please elaborate why initializing `i` with this value? Also, why breaking on line 690?

Reply:
If viewport is 0, you're adding it to all viewport, and since they're numbered incrementally their id numbers will be `{1, ..., #existing_viewports}`. If viewport is bigger than 0, then you know the user is specifying the specific viewport id to add the text.

So this means that the ids you will be addressing are either from [1, #viewports] or simply the one specified in viewport, as long as its bigger then 0.

Regarding the break, you only cycle all possible ids when viewport = 0. If the user specified a single id, only that id is to be checked, that's why you break after a single iteration.